src/vs-bicep/Bicep.VSLanguageServerClient/ProcessTracker/WindowEventsListener.cs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Bicep.VSLanguageServerClient.ProcessTracker
+{
+    public class WindowEventsListener : IWindowEventsListener
+    {
+        private WindowEventHandler _handler;
+        private GCHandle _handlerGCPin;
+        private IntPtr _hWinEventHook;
+
+        public WindowEventsListener(
+            WindowsSystemEvents min = WindowsSystemEvents.EventMin,
+            WindowsSystemEvents max = WindowsSystemEvents.EventMax)
+        {
+            _handler = new WindowEventHandler(InternalSystemEventHandler);
+            _handlerGCPin = GCHandle.Alloc(_handler);
+
+            _hWinEventHook = NativeMethods.SetWinEventHook(
+                min,
+                max,
+                IntPtr.Zero,
+                _handler,
+                0,
+                0,
+                (uint)(WinEventHookFlags.OutOfContext | WinEventHookFlags.SkipOwnProcess));
+        }
+
+        public event EventHandler<WindowEventArgs>? SystemEvent;
+
+        private void InternalSystemEventHandler(
+            IntPtr hWinEventHook,
+            WindowsSystemEvents anEvent,
+            IntPtr hwnd,
+            int idObject,
+            int idChild,
+            uint dwEventThread,
+            uint dwmsEventTime)
+        {
+            SystemEvent?.Invoke(this, new WindowEventArgs(anEvent, hwnd));
+        }
+
+        ~WindowEventsListener()
+        {
+            Dispose(false);
+        }
+
+        #region IDisposable Members
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        #endregion
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_hWinEventHook != IntPtr.Zero)
+            {
+                NativeMethods.UnhookWinEvent(_hWinEventHook);
+                _hWinEventHook = IntPtr.Zero;
+            }
+
+            if (disposing)
+            {
+                if (_handlerGCPin.IsAllocated)
+                {
+                    _handlerGCPin.Free();
+                }
+
+                _handler = null!;
+            }
+        }
+    }
+
+    public delegate void WindowEventHandler(
+        IntPtr hWinEventHook,
+        WindowsSystemEvents anEvent,
+        IntPtr hwnd,
+        int idObject,
+        int idChild,
+        uint dwEventThread,
+        uint dwmsEventTime);
+
+    public class WindowEventArgs : EventArgs
+    {
+        public WindowEventArgs(WindowsSystemEvents anEvent, IntPtr windowHandle)
+        {
+            SystemEvent = anEvent;
+            WindowHandle = windowHandle;
+        }
+
+        public WindowsSystemEvents SystemEvent { get; }
+        public IntPtr WindowHandle { get; }
+    }
```

Comment:
Can we move these to a different file?

Reply:
Same for the enum below.

Reply:
Sure, updated here:
https://github.com/Azure/bicep/commit/b61d201d404726e487e39be7aa1791364385e2d2